### ISSUE TITLE
Mounting local user to docker

### ISSUE THREAD

**Issue by ilinojkovic** [microsoft/vscode-remote-release#551]
This is to some extent connected to my previous question in [#550](https://github.com/microsoft/vscode-remote-release/issues/550).

My end goal is to have a local docker container used only as a dev environment, meaning I can build and run my code through docker, also display some results [#550](https://github.com/microsoft/vscode-remote-release/issues/550), while the code itself is accessible and editable from my local machine.

To achieve this, I would like to mount my local user (which also has sudo privileges) to docker, so when files are created I can easily open them either in docker or from local machine. Before opting to use this extension, I used this script to run the container:

```
#!/bin/bash
xhost +local:
VERSION=$"3"
docker run -it --net=host \
  -u=$(id -u) \
  -e DISPLAY=$DISPLAY \
  -e QT_GRAPHICSSYSTEM=native \
  -e CONTAINER_NAME=kinetic_v${VERSION} \
  -e USER=$USER \
  --workdir=/absolute_path_to/workspace \
  -v "/tmp/.X11-unix:/tmp/.X11-unix" \
  -v "/etc/group:/etc/group:ro" \
  -v "/etc/passwd:/etc/passwd:ro" \
  -v "/etc/shadow:/etc/shadow:ro" \
  -v "/etc/sudoers.d:/etc/sudoers.d:ro" \
  -v "/home/$USER/:/home/$USER/" \
  --device=/dev/dri:/dev/dri \
  --name=kinetic_v${VERSION} \
  ilinojkovic/ros:kinetic_v${VERSION}
```

This provided me with what I intended, however, I tried to translate this into `devcontainer.json`, but with little success:

```
{
	"name": "kinetic_v5",
	"context": "..",
	"dockerFile": "Dockerfile",
	"workspaceFolder": "/workspace",
	"runArgs": [
		"--net", "host",
		"-e", "DISPLAY=${env:DISPLAY}",
		"-e", "QT_GRAPHICSSYSTEM=native",
		"-e", "CONTAINER_NAME=kinetic_v5",
		"-v", "/tmp/.X11-unix:/tmp/.X11-unix",
		"-v", "/etc/group:/etc/group:ro",
		"-v", "/etc/passwd:/etc/passwd:ro",
		"-v", "/etc/shadow:/etc/shadow:ro",
		"-v", "/etc/sudoers.d:/etc/sudoers.d:ro",
		"-v", "/home/${env:USER}/:/home/${env:USER}/",
		"-u", "1000",            // idk how to run $(id -u) here
		"-e", "USER=${env:USER}",
		"--device=/dev/dri:/dev/dri",
		"--name=kinetic_v5",
	],
	"extensions": [
		"ms-python.python"
	]
}
```

In the end I get stuck with the error:
```
mkdir: cannot create directory '/root': Permission denied
```
since vscode tries to add all the config files to `/root` dir. Can this path be changed to a home folder? Would only changing this resolve the issue? And if not, what is your advice to achieve this setup?

Setup details:
- OS: linux
- Product: Visual Studio Code - Insiders
- Product Version: 1.35.0-insider
- Language: en

**Comment by chrmarti** [comment]
Could you try changing the `HOME` environment variable to the user's home folder? That's where the VS Code Server will be installed to.

**Comment by ilinojkovic** [comment]
Thanks for the reply. Adding `"-e", "HOME=${env:HOME}"` to `runArgs` did resolve the issue with `/root` access, however, I'm now prompted with a different one:
```
OCI runtime exec failed: exec failed: container_linux.go:345: starting container process caused "chdir to cwd (\"/home/ilijas/.vscode-server-insiders/bin/a5536b8f5a16a10d859f3dec1e59701671bf069e\") set in config.json failed: permission denied": unknown
``` 
It is strange that it got a `permission denied` since the folder `home/ilijas/.vscode-server-insiders/bin/a5536b8f5a16a10d859f3dec1e59701671bf069e` and it's contents are created with appropriate user without any issues.

**Comment by ilinojkovic** [comment]
This is the full output:
```
Setting up container for folder: /home/ilijas/path_to/workspace
Run: docker build -f /home/ilijas/path_to/workspace/.devcontainer/Dockerfile -t vsc-workspace-e3de0c09162a91206074b2e928584a75 /home/ilijas/path_to/workspace
Sending build context to Docker daemon  1.395GB
Step 1/5 : FROM docker.is.localnet:5000/amd/official:16.04
 ---> 374edab0eb1d
Step 2/5 : RUN apt-get update &&     apt-get install -y zsh     fonts-powerline     locales     && locale-gen en_US.UTF-8
 ---> Using cache
 ---> b7e6eb0d1d78
Step 3/5 : RUN pip2 install Cython gym
 ---> Using cache
 ---> 48df31225560
Step 4/5 : RUN wget https://github.com/robbyrussell/oh-my-zsh/raw/master/tools/install.sh -O - | zsh || true
 ---> Using cache
 ---> deb64b23c110
Step 5/5 : CMD ["zsh"]
 ---> Using cache
 ---> ef2343fe5c3b
Successfully built ef2343fe5c3b
Successfully tagged vsc-workspace-e3de0c09162a91206074b2e928584a75:latest
Run: docker run -a STDOUT -a STDERR --mount type=bind,source=/home/ilijas/path_to/workspace,target=/workspaces/workspace -v /home/ilijas/.gitconfig:/root/.gitconfig -l vsch.quality=insider -l vsch.local.folder=/home/ilijas/path_to/workspace -l vsch.remote.devPort=0 --net host -e HOME=/home/ilijas -e DISPLAY=:0 -e QT_GRAPHICSSYSTEM=native -e CONTAINER_NAME=kinetic_v5 -v /tmp/.X11-unix:/tmp/.X11-unix -v /etc/group:/etc/group:ro -v /etc/passwd:/etc/passwd:ro -v /etc/shadow:/etc/shadow:ro -v /etc/sudoers.d:/etc/sudoers.d:ro -v /home/ilijas/:/home/ilijas/ -u 1000 -e USER=ilijas --device=/dev/dri:/dev/dri --name=kinetic_v5 vsc-workspace-e3de0c09162a91206074b2e928584a75 /bin/sh -c echo Container started ;  while sleep 1; do :; done
Container started
Run: docker exec bbc1d49a572f8968b2dc17aac1bfc560e76feb6e5d91587c2892ccfe5b2228a7 test -d /home/ilijas/.vscode-server-insiders
Run: docker exec bbc1d49a572f8968b2dc17aac1bfc560e76feb6e5d91587c2892ccfe5b2228a7 /bin/sh -c set -o noclobber ; mkdir -p '/home/ilijas/.vscode-server-insiders/data/Machine' && { > '/home/ilijas/.vscode-server-insiders/data/Machine/.writeMachineSettingsMarker' ; } 2> /dev/null
Run: docker exec bbc1d49a572f8968b2dc17aac1bfc560e76feb6e5d91587c2892ccfe5b2228a7 /bin/sh -c (cat /etc/os-release || cat /usr/lib/os-release) 2>/dev/null
NAME="Ubuntu"
VERSION="16.04.6 LTS (Xenial Xerus)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 16.04.6 LTS"
VERSION_ID="16.04"
HOME_URL="http://www.ubuntu.com/"
SUPPORT_URL="http://help.ubuntu.com/"
BUG_REPORT_URL="http://bugs.launchpad.net/ubuntu/"
VERSION_CODENAME=xenial
UBUNTU_CODENAME=xenial
Run: docker exec bbc1d49a572f8968b2dc17aac1bfc560e76feb6e5d91587c2892ccfe5b2228a7 test -d /home/ilijas/.vscode-server-insiders/bin/a5536b8f5a16a10d859f3dec1e59701671bf069e
Run: docker exec bbc1d49a572f8968b2dc17aac1bfc560e76feb6e5d91587c2892ccfe5b2228a7 test -d /home/ilijas/.vscode-server-insiders/extensions
Run: docker exec bbc1d49a572f8968b2dc17aac1bfc560e76feb6e5d91587c2892ccfe5b2228a7 /bin/sh -c set -o noclobber ; mkdir -p '/home/ilijas/.vscode-server-insiders/data/Machine' && { > '/home/ilijas/.vscode-server-insiders/data/Machine/.postCreateCommandMarker' ; } 2> /dev/null
Run: docker exec -w /home/ilijas/.vscode-server-insiders/bin/a5536b8f5a16a10d859f3dec1e59701671bf069e -u 0 bbc1d49a572f8968b2dc17aac1bfc560e76feb6e5d91587c2892ccfe5b2228a7 /bin/sh -c command -v git >/dev/null 2>&1 && git config --system credential.helper '!f() { command -v code-insiders >/dev/null 2>&1 && code-insiders --gitCredential $*; }; f' || true
OCI runtime exec failed: exec failed: container_linux.go:345: starting container process caused "chdir to cwd (\"/home/ilijas/.vscode-server-insiders/bin/a5536b8f5a16a10d859f3dec1e59701671bf069e\") set in config.json failed: permission denied": unknown
Command failed: docker exec -w /home/ilijas/.vscode-server-insiders/bin/a5536b8f5a16a10d859f3dec1e59701671bf069e -u 0 bbc1d49a572f8968b2dc17aac1bfc560e76feb6e5d91587c2892ccfe5b2228a7 /bin/sh -c command -v git >/dev/null 2>&1 && git config --system credential.helper '!f() { command -v code-insiders >/dev/null 2>&1 && code-insiders --gitCredential $*; }; f' || true
```

**Comment by chrmarti** [comment]
The failing command runs as root. Does root have access to the `/home/ilijas/.vscode-server-insiders/bin/a5536b8f5a16a10d859f3dec1e59701671bf069e` folder?

**Comment by ilinojkovic** [comment]
Yes, sudo/root can access that folder: https://i.imgur.com/EiWOM5z.png

**Comment by chrmarti** [comment]
Is that inside the container? I see you have your home folder mounted. Maybe the root user of the container cannot access that folder. Can you check that?

**Comment by ilinojkovic** [comment]
In the previous case I checked from the local machine, and here's the output of some of the commands run inside docker:
```
carbon ➜ / (kinetic_v5) echo $USER                                
ilijas
carbon ➜ / (kinetic_v5) echo $HOME                                
/home/ilijas
carbon ➜ / (kinetic_v5) id -u                                     
1000
carbon ➜ / (kinetic_v5) ll home/ilijas/.vscode-server-insiders/bin
total 4.0K
drwxr-xr-x 6 ilijas ilijas 4.0K Jun  4 13:29 a5536b8f5a16a10d859f3dec1e59701671bf069e
carbon ➜ / (kinetic_v5) ll home/ilijas/.vscode-server-insiders/bin/a5536b8f5a16a10d859f3dec1e59701671bf069e 
total 36M
-rw-r--r--   1 ilijas ilijas  14K May 20 05:03 LICENSE
drwxr-xr-x   2 ilijas ilijas 4.0K May 20 05:22 bin
drwxr-xr-x  27 ilijas ilijas 4.0K May 20 05:24 extensions
-rwxr-xr-x   1 ilijas ilijas  36M May 20 05:22 node
drwxr-xr-x 152 ilijas ilijas 4.0K May 20 05:23 node_modules
drwxr-xr-x   3 ilijas ilijas 4.0K May 20 05:22 out
-rw-r--r--   1 ilijas ilijas  837 May 20 05:03 package.json
-rw-r--r--   1 ilijas ilijas  18K May 20 05:03 product.json
-rwxr-xr-x   1 ilijas ilijas  232 May 20 05:03 server.sh
carbon ➜ / (kinetic_v5) cat /etc/passwd                           
root:x:0:0::/root:/usr/bin/zsh
nobody:x:65534:65534:Nobody:/:/sbin/nologin
dbus:x:81:81:System Message Bus:/:/sbin/nologin
bin:x:1:1::/:/sbin/nologin
daemon:x:2:2::/:/sbin/nologin
mail:x:8:12::/var/spool/mail:/sbin/nologin
ftp:x:14:11::/srv/ftp:/sbin/nologin
http:x:33:33::/srv/http:/sbin/nologin
systemd-journal-remote:x:982:982:systemd Journal Remote:/:/sbin/nologin
systemd-network:x:981:981:systemd Network Management:/:/sbin/nologin
systemd-resolve:x:980:980:systemd Resolver:/:/sbin/nologin
systemd-timesync:x:979:979:systemd Time Synchronization:/:/sbin/nologin
systemd-coredump:x:978:978:systemd Core Dumper:/:/sbin/nologin
uuidd:x:68:68::/:/sbin/nologin
dnsmasq:x:977:977:dnsmasq daemon:/:/sbin/nologin
rpc:x:32:32:Rpcbind Daemon:/var/lib/rpcbind:/sbin/nologin
avahi:x:975:975:Avahi mDNS/DNS-SD daemon:/:/sbin/nologin
colord:x:974:974:Color management daemon:/var/lib/colord:/sbin/nologin
cups:x:209:209:cups helper user:/:/sbin/nologin
geoclue:x:973:973:Geoinformation service:/var/lib/geoclue:/sbin/nologin
git:x:972:972:git daemon user:/:/usr/bin/git-shell
nm-openconnect:x:971:971:NetworkManager OpenConnect:/:/sbin/nologin
nm-openvpn:x:970:970:NetworkManager OpenVPN:/:/sbin/nologin
ntp:x:87:87:Network Time Protocol:/var/lib/ntp:/bin/false
polkitd:x:102:102:PolicyKit daemon:/:/sbin/nologin
rtkit:x:133:133:RealtimeKit:/proc:/sbin/nologin
sddm:x:969:969:Simple Desktop Display Manager:/var/lib/sddm:/sbin/nologin
usbmux:x:140:140:usbmux user:/:/sbin/nologin
ilijas:x:1000:1000:Ilija Stanojkovic:/home/ilijas:/usr/bin/zsh
test:x:1001:1001::/home/test:/bin/bash
test2:x:1002:1002::/home/test2:/bin/bash
flatpak:x:967:967:Flatpak system helper:/:/sbin/nologin
sndiod:x:966:966::/var/lib/sndiod:/bin/false
```
Everything seems to be in order. There exists a root user, but the user after login is `ilijas` which is correct. Maybe in the docker setup it switches to `root` and back, but still it has access to all of the files and folders. The odd thing that I can only notice is that even though the folder `a5536b8f5a16a10d859f3dec1e59701671bf069e` is created yesterday, the contents of it are created a couple of weeks back. Idk if this is expected or not.

**Comment by chrmarti** [comment]
The timestamps on the files are from the archive they were extracted from.

What do you get for (replace the container id if that has changed since you posted the log):
```
docker exec -u 0 -w /home/ilijas/.vscode-server-insiders/bin/a5536b8f5a16a10d859f3dec1e59701671bf069e bbc1d49a ls
```

I would expect the same error to show. Does it work when removing `-u 0` from the command?

**Comment by ilinojkovic** [comment]
Hi @chrmarti, sorry for the late response.

Yes, the behavior is exactly as you described. I presume the part `-u 0` switches to the root user. Without that, it prints the contents of the folder. Nonetheless, this seems to be strange, since the root does have access to the files and folders when I check.

**Comment by chrmarti** [comment]
The container's root user might not have the expected permissions on the mounted folder. I will change it such that this command runs in `/` of the container as its working dir, that should always be available to root.

**Comment by ilinojkovic** [comment]
Super, looking forward to the change. Many thanks!

I would just like express my satisfaction with the product (both vscode and this extension), but imo, what makes it special is the decision to do all the development out in the open. Being open source instills trust, but also makes it easier for community to contribute in whichever way possible. I've been using vscode for a bit more than a month, and apart from a couple of hiccups, it's been quite a smooth ride. Keep up the good work! :smiley:

**Comment by chrmarti** [comment]
Awesome, thanks @ilinojkovic! And thanks for taking the time to follow-up on this issue! :)

**Comment by aeschli** [comment]
@ilinojkovic Can you verify that the fix worked for you?
You need vscode insiders for that and the latest Remote Container extension.

**Comment by ilinojkovic** [comment]
So I updated vscode insiders and I'm running Remote Development extension 0.15.0 which seems to be the latest official one, and I get the same error as before.

**Comment by chrmarti** [comment]
Are you using VS Code Insiders? (That gets the latest version of Remote Containers with the fix.)

**Comment by ilinojkovic** [comment]
Yes, I updated the package from command line like any other. Is there another way for insiders version?

This is the output from Help->About:
Version: 1.35.0-insider
Commit: a5536b8f5a16a10d859f3dec1e59701671bf069e
Date: 2019-05-20T05:15:35.934Z
Electron: 3.1.8
Chrome: 66.0.3359.181
Node.js: 10.2.0
V8: 6.6.346.32
OS: Linux x64 4.19.49-1-MANJARO

**Comment by chrmarti** [comment]
That is an older version of Insiders. Maybe you turned automatic updates off? Please update to (or download and install) the latest version of Insiders (currently from 2019-06-28) and try again.

**Comment by ilinojkovic** [comment]
Hi @chrmarti. I'm using Manjaro, an arch based linux, which is not officially supported. I now see that the maintainer doesn't increment the version as regularly as insider versions change, so I'll see if this changes during the week. Otherwise I'll probably need to build the thing myself. Hopefully it's not a problem if I get back to you by Monday.

**Comment by ilinojkovic** [comment]
Hello @chrmarti, I updated vscode to:

Version: 1.37.0-insider
Commit: 7e03eb1a1fce3beb6ccea5f3b8f828e01940be61
Date: 2019-07-05T19:16:48.439Z
Electron: 4.2.5
Chrome: 69.0.3497.128
Node.js: 10.11.0
V8: 6.9.427.31-electron.0
OS: Linux x64 4.19.56-1-MANJARO

Good news - I rebuilt the image, and now my local user is perfectly mounted. As a bonus the issue #550 also got resolved as well. :smile: 

Thanks again for all the help! Awesome work from you and your team.
